File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    int num = 0;
    
    public void check(List<Boolean> visited, int[][] wires, int startNum) {
        for (int[] w : wires) {
            int nextNum;
            if (w[0] == startNum && !visited.get(w[1] - 1)) {
                nextNum = w[1];
            } else if (w[1] == startNum && !visited.get(w[0] - 1)) {
                nextNum = w[0];
            } else {
                continue;
            }
            
            num += 1;
            visited.set(nextNum - 1, true);
            check(visited, wires, nextNum);
            visited.set(nextNum - 1, false);
        }
    }
    
    public int solution(int n, int[][] wires) {
        List<Boolean> visited = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            visited.add(false);
        }
        
        int answer = -1;
        int one, two;
        for (int[] w: wires) {
            visited.set(w[0] - 1, true);
            visited.set(w[1] - 1, true);
            
            num = 1;
            check(visited, wires, w[0]);
            one = num;
            
            num = 1;
            check(visited, wires, w[1]);
            two = num;
            
            if (answer == -1) answer = Math.abs(one - two);
            else if (answer > Math.abs(one - two)) answer = Math.abs(one - two);
            
            visited.set(w[0] - 1, false);
            visited.set(w[1] - 1, false);
        }
        
        return answer;
    }
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
def solution(n, wires):
    num = 0
    def check(visited, start_num):
        nonlocal num
        for w in wires:
            if start_num not in w:
                continue
            if start_num == w[0] and not visited[w[1] - 1]:
                next_num = w[1] 
            elif start_num == w[1] and not visited[w[0] - 1]:
                next_num = w[0]
            else:
                continue
            num += 1
            visited[next_num - 1] = True
            check(visited, next_num)
            visited[next_num - 1] = False
            
    answer = -1
    visited = [False] * n
    for w in wires:
        visited[w[0] - 1] = True
        visited[w[1] - 1] = True
        num = 1
        check(visited, w[0])
        one = num
        num = 1
        check(visited, w[1])
        two = num
        if answer == -1: 
            answer = abs(one - two)
        elif answer > abs(one - two):
            answer = abs(one - two)
        visited[w[0] - 1] = False
        visited[w[1] - 1] = False
        
    return answer